Grid serialization issue with AbstractDataProvider.addDataProviderListener

Env: Flow 14.0.0-rc7, Java 1.8.0_122, Ubuntu 19.04

When trying to serialize a `VaadinSession` that contains a `Grid` an exception is thrown with the following cause

```
Caused by: java.lang.ClassCastException: java.lang.invoke.SerializedLambda cannot be cast to com.vaadin.flow.function.SerializableConsumer
        at com.vaadin.flow.data.provider.AbstractDataProvider.$deserializeLambda$(AbstractDataProvider.java:41)
        ... 781 more
```

The problem can be replicated on `bookstore-starter-flow` adding adding a `RequestHandler` that performs the serialization of the current vaadin session (`serializeAndDeserialize` method is taken from Vaadin tests) .

```java
public class SerializationTestListener implements VaadinServiceInitListener {
    @Override
    public void serviceInit(ServiceInitEvent initEvent) {
        initEvent.addRequestHandler((session, request, response) -> {
            try {
                serializeAndDeserialize(session);
            } catch (Throwable t) {
                t.printStackTrace();
            }
            return false;
        });
    }

    public <T> T serializeAndDeserialize(T instance) throws Throwable {
        ByteArrayOutputStream bs = new ByteArrayOutputStream();
        ObjectOutputStream out = new ObjectOutputStream(bs);
        out.writeObject(instance);
        byte[] data = bs.toByteArray();
        ObjectInputStream in = new ObjectInputStream(
            new ByteArrayInputStream(data));

        @SuppressWarnings("unchecked")
        T readObject = (T) in.readObject();

        return readObject;
    }
}
```

I've pushed the branch `serialization-error` on my bookstore starter fork with the above configuration and some other fix always related to serialization; you can find it here: 
https://github.com/mcollovati/bookstore-starter-flow/tree/serialization-error

The problem seems to be similar to the one reported in #5973 

Using an anonymous class  for the `SerializableConsumer` instead of a method reference in `AbstractDataProvider.addDataProviderListener` seems to solve the problem

```java
     public Registration addDataProviderListener(
             DataProviderListener<T> listener) {
        return addListener(DataChangeEvent.class, new SerializableConsumer<DataChangeEvent>() {
            @Override
            public void accept(DataChangeEvent dataChangeEvent) {
                listener.onDataChange(dataChangeEvent);
            }
        });
        // Original implementation
        // return addListener(DataChangeEvent.class, listener::onDataChange);
     }
```

--- a/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractDataProvider.java
+++ b/flow-data/src/main/java/com/vaadin/flow/data/provider/AbstractDataProvider.java
@@ -45,7 +45,16 @@ public abstract class AbstractDataProvider<T, F> implements DataProvider<T, F> {
     @Override
     public Registration addDataProviderListener(
             DataProviderListener<T> listener) {
-        return addListener(DataChangeEvent.class, listener::onDataChange);
+        // Using an anonymous class instead of lambda or method reference to prevent potential
+        // self reference serialization issues when clients holds a reference
+        // to the Registration instance returned by this method
+        SerializableConsumer<DataChangeEvent> consumer = new SerializableConsumer<DataChangeEvent>() {
+            @Override
+            public void accept(DataChangeEvent dataChangeEvent) {
+                listener.onDataChange(dataChangeEvent);
+            }
+        };
+        return addListener(DataChangeEvent.class, consumer);
     }
 
     @Override

--- a/flow-data/src/test/java/com/vaadin/flow/tests/data/DataSerializableTest.java
+++ b/flow-data/src/test/java/com/vaadin/flow/tests/data/DataSerializableTest.java
@@ -1,6 +1,58 @@
 package com.vaadin.flow.tests.data;
 
+import java.io.Serializable;
+import java.util.Collections;
+
+import com.vaadin.flow.data.binder.HasDataProvider;
+import com.vaadin.flow.data.provider.DataProvider;
+import com.vaadin.flow.data.provider.ListDataProvider;
+import com.vaadin.flow.shared.Registration;
 import com.vaadin.flow.testutil.ClassesSerializableTest;
+import org.junit.Test;
+
+import static org.junit.Assert.assertNotNull;
 
 public class DataSerializableTest extends ClassesSerializableTest {
+
+    /*
+     * AbstractDataProvider.addDataProviderListener may return a Registration instance
+     * that is not deserializable due to self references.
+     * This happens for example if the dataprovider, member of a component,
+     * is used to add a com.vaadin.flow.data.provider.DataProviderListener
+     * into an inner component; the resulting Registration handles a reference
+     * to the dataprovider itself that is already referenced by the outer component
+     */
+    @Test
+    public void selfReferenceSerialization() throws Throwable {
+        Outer outer = new Outer();
+        Outer out = serializeAndDeserialize(outer);
+        assertNotNull(out);
+    }
+
+    static class Inner implements HasDataProvider<Object>, Serializable {
+
+        private Registration registration;
+
+        @Override
+        public void setDataProvider(DataProvider<Object, ?> dataProvider) {
+            if (registration != null) {
+                registration.remove();
+            }
+            registration = dataProvider.addDataProviderListener(event -> onDataProviderChange());
+        }
+
+        void onDataProviderChange() {
+
+        }
+
+    }
+
+    static class Outer implements Serializable {
+        private final ListDataProvider<Object> dataProvider = new ListDataProvider<>(Collections.emptyList());
+        private final Inner inner = new Inner();
+
+        public Outer() {
+            inner.setDataProvider(dataProvider);
+        }
+    }
 }

